### PR TITLE
[ENH] optimize metric state updates to prevent O(N^2) reallocation

### PR DESCRIPTION
**Pull request by andersendsa**
#### What does this implement/fix? Explain your changes.
Initialize states with empty lists instead of dummy tensors for reduction='none'. Append to list during `update` and only do a single `torch.cat` inside `compute`. This dramatically speeds up long evaluation loops for multi-horizon metrics.

#### PR checklist

- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.
- [ ] Added/modified tests
- [x] Used pre-commit hooks when committing to ensure that code is compliant with hooks. Install hooks with `pre-commit install`.
  To run hooks independent of commit, execute `pre-commit run --all-files`

### CODE REVIEW

**Review by phoeenniixx — CHANGES_REQUESTED** [review]
Thanks! Can you please share an example how this helps?
Here you check if `losses` is a `list`, when does that happen?

File: pytorch_forecasting/metrics/base_metrics/_base_metrics.py [hunk 1]
```diff
@@ -897,12 +897,17 @@ def update(self, y_pred, target):
     def _update_losses_and_lengths(self, losses: torch.Tensor, lengths: torch.Tensor):
         losses = self.mask_losses(losses, lengths)
         if self.reduction == "none":
-            if self.losses.ndim == 0:
-                self.losses = losses
-                self.lengths = lengths
+            if isinstance(self.losses, list):
+                self.losses.append(losses)
```
Comment (Shriprasad-P) [inline pytorch_forecasting/metrics/base_metrics/_base_metrics.py:901]: [P2] Add a multi-update regression test for this new list-accumulation path. Existing coverage invokes `MAE(reduction="none")` only once via `forward`; the parametrized `test_reduction_modes` leaves the metric at its default `mean`. Nothing asserts that multiple direct `update()` calls are appended and concatenated correctly, or that `reset()` clears the list state. Please add a direct `MAE(reduction="none")` test with uneven batch sizes that checks `compute()` against the expected concatenation, then `reset() -> update() -> compute()`. This is the core changed lifecycle and would catch a regression that assigns rather than appends.

**Review by Shriprasad-P — CHANGES_REQUESTED** [review]
I verified numerical parity, state lifecycle, TorchMetrics list-state behavior, CPU movement, two-rank Gloo synchronization, and the performance improvement. The requested regression coverage below is the only blocking concern.

**Review by Shriprasad-P — CHANGES_REQUESTED** [review]
Could we add a direct regression test for the new multi-update reduction="none" lifecycle here?
The current tests do not appear to cover multiple direct update() calls being accumulated into the list state and then concatenated in compute(), nor do they verify that reset() clears that accumulated state.

A focused MAE(reduction="none") test with uneven batch sizes that checks compute() against the expected concatenated result, followed by reset() -> update() -> compute(), would protect the core behavior introduced by this change.

**Review by Shriprasad-P — COMMENTED** [review]
## Review

Accumulating `reduction="none"` updates in a list and concatenating once in `compute()` is a clean way to avoid repeated `torch.cat` O(N²) reallocations.

Fallback path when state is already a tensor is sensible for distributed concat behavior. Worth a quick check that Lightning’s state sync still works with list defaults, but the approach looks sound.
